src/commcare_cloud/ansible/group_vars/all.yml

```diff
@@ -5,7 +5,6 @@ code_home: "{{ www_home }}/current"
 log_home: "{{ www_home }}/log"
 legacy_service_home: "{{ code_home }}/services"
 service_home: "{{ www_home }}/services"
-virtualenv_home: "{{ code_home }}/python_env"
 py3_virtualenv_home: "{{ code_home }}/python_env-3.6"
```

Comment:
You're talking about changing `{{ code_home }}/python_env-3.6` to `{{ code_home }}/python_env` (or `{{ code_home }}/pyenv`), right?

I think my preference is not to reuse `python_env` to now mean the python3 env, because if people report issues on the forum it's clearer what's going on if we can distinguish paths that have to do with the old python2 env versus the new python3 env, and it allows us to have the longest tail of support not for _running_ python2, but for _upgrading from_ a python2 environment to a python3 environment. If we replace the current name, then that creates some confusion in the upgrade process, and may (I think currently would) either error or not actual result in an upgraded environment. If the virtualenv upgrade process had been an in-place one using the same name would be the obvious choice, but since the upgrade process requires creating a new one, I'd rather keep the names separate.

That doesn't rule out using a name where the difference doesn't semantically convey the difference between the two (like `python_env` for lingering python 2 and `pyenv` for python 3), but there's of course something a little awkward about that.

Looking ahead, I can easily see us upgrading to python3.7 in the future, and it my guess is that again we'd choose to do it by creating a new virtualenv that we build from scratch, rather than updating the current virtualenv in place. It'd be cool if whatever scheme we use also solves that future problem in much the same way.

So, given all that, I might suggest the following:
- keep the virtualenv name as it is: `python_env-3.6`
- name future virtualenvs according to the same scheme, e.g. `python_env-3.7`
- create a symlink `pyenv -> python_env-3.6` for use by developers to protect muscle memory
- In our automation/code, always refer to envs by their real path and not the symlink path

This last point I think is important unless we thoroughly test all the potential limitations of a virtualenv path that contains a symlink. For example if that were to affect virtualenv-clone's ability to do its stuff cleanly, that would of course be an issue. (I think we've had problems with symlinks before, but I think that was symlinks inside the virtualenv or something, not the path _to_ the virtualenv.) If that's all well tested, then using the opposite convention—always referring to the current python env as pyenv—could be preferable. Even then though it has some drawbacks. For example, during the next python update we undergo, we will want to make sure we're cloning the right env with the right version of python. If there's some hidden context/state wrapped up in what `pyenv` is actually pointing to, this could get confusing and error prone.

Reply:
I like the suggestions.

> In our automation/code, always refer to envs by their real path and not the symlink path

This seems fine as long as we can keep the number of places where we reference it in code to a minimum so it is not a burdensome task to update all of the references when we upgrade.

One question that comes to mind is "Why haven't we had to deal with this in the past?" I think the answer is that Python 2.7 was supported forever, so we do not have a very sturdy upgrade process in place for the Python interpreter.

Reply:
Are you planning to implement all of those suggestions in this PR, or some of them (ex. the symlink to `pyenv`) in a separate PR?

Reply:
If it's okay, maybe I'd prefer adding the `pyenv` in a separate PR. Just realized it requires some testing to make sure that if I set it up in ansible a deploy doesn't "undo" it (by making not doing it in the new release) and/or decide to only do it in fab. Not a huge amount of work, but I just want to not delay this since it touches a lot of code and is ripe for potential conflicts

Reply:
👍 That's fine